### PR TITLE
View and modify node connection settings individually

### LINKED ISSUE #101: Display connection settings and macaroons in settings page
At the moment, the only option is 'Reset password and connection settings'. It'd be nice to:

- see the connection settings without having to reset them
- be able to update the REST URL, macaroons and password (#35) separately.

### PR DESCRIPTION
**Pull request by jamaljsr**
Closes #101 and #35

### Description

Adds fields on the settings page to view the current connection settings (REST URL & macaroons). Added the ability to change the URL, admin/readonly macaroons, and the password individually.

### Steps to Test

**Node REST URL**
1. Navigate to the Settings screen
2. Click on the edit icon for the field labeled *REST API URL*
3. Enter a valid URL and confirm the it completes successfully
4. Enter a invalid URL (ex: https://localhost:1234) and confirm an error message is displayed

*Note: It is currently possible to enter a URL for a different node which renders the current macaroons invalid. Should we prevent this behavior by checking macaroons in addition to the REST URL?*

**Admin/Readonly Macaroons**
1. Navigate to the Settings screen
2. Click on the edit icon for either macaroon field. They both open the same form.
3. Modify the form by uploading a new macaroon for one or both fields. You can also switch to the HEX input fields. Both fields are optional because the form is initialized with the current macaroon values already set.
4. Save the form and confirm the expected result

**Change Password**
1. Navigate to the Settings screen
2. Click on the *Change Password* button
3. Enter your current password in the modal if it has not been already saved for the session
4. Enter in a new password then click continue
5. Close and reopen the extension window 
6. Click on the `Deposit` button to trigger a password prompt
7. Type in your new password and confirm you are able to view the deposit QR code

### CODE REVIEW

File: src/app/components/Settings/NodeSettings.tsx [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+          // isChangingPassword true -> false and password changed
+          message.success('Password Updated', 2);
+    }
+    const macaroonsChanged = adminMacaroon !== nextProps.adminMacaroon
+      || readonlyMacaroon !== nextProps.readonlyMacaroon;
+    if (isUpdatingMacaroons && macaroonsChanged) {
+        // isUpdatingMacaroons is true and a macaroon changed
+        message.success('Macaroons Updated', 2);
+    }
+  }
```
Comment (jamaljsr) [inline src/app/components/Settings/NodeSettings.tsx]: I don't like this 💩 code at all. I have some thoughts on how to improve it throughout the app using an alerts reducer, but I'll save that discussion for a separate issue/PR.

File: src/app/style.less [hunk 2]
```diff
@@ -50,3 +50,12 @@ body {
     }
   }
 }
+
+.ant-drawer-body {
+  height: calc(100vh - @drawer-header-height);
```
Comment (jamaljsr) [inline src/app/style.less]: This allows the drawer to be full height so that the password disclaimer is positioned at the bottom.

File: src/app/style.less [hunk 3]
```diff
@@ -50,3 +50,12 @@ body {
...
+}
+
+// ensure modals display above drawers
+.ant-modal-mask, .ant-modal-wrap {
+  z-index: 1001;
```
Comment (jamaljsr) [inline src/app/style.less]: This fixes an issue where the password prompt modal would be displayed behind the drawer if any modal was once displayed since the page loaded.

image

File: src/app/modules/node/sagas.ts [hunk 4]
```diff
@@ -92,6 +100,80 @@ export function* handleCheckAuth(action: ReturnType<typeof actions.checkAuth>):
...
+    
+    // passowrd is needed to decrypt the admin macaroon
+    yield call(requirePassword);
+
+    // get current macaroons from state as its needed to store the new url
+    const { readonlyMacaroon } = yield select(selectSyncedUnencryptedNodeState);
+    const { adminMacaroon } = yield select(selectSyncedEncryptedNodeState);
+
+    // connect to the url to test if it's working
+    yield put(actions.checkNode(newUrl));
```
Comment (wbobeirne) [inline src/app/modules/node/sagas.ts]: This action sets `url` to null, and if the connection fails, it leaves it null in the reducer. This is fixed on a new session if it doesn't get synced, but if it did, it would bork your session.

Comment (wbobeirne) [reply src/app/modules/node/sagas.ts]: screen shot 2019-01-10 at 11 54 14 pm

Note the empty input on the left

File: src/app/components/Settings/NodeSettings.tsx [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+    } = this.props;
+
+    if (isNodeChecked !== nextProps.isNodeChecked && nextProps.isNodeChecked) {
+          // isNodeChecked false -> true
+          message.success(`Connected to ${nextProps.url}`, 2);
+    }
+    if (isChangingPassword !== nextProps.isChangingPassword && !isChangingPassword
+        && password !== nextProps.password) {
+          // isChangingPassword true -> false and password changed
+          message.success('Password Updated', 2);
```
Comment (wbobeirne) [inline src/app/components/Settings/NodeSettings.tsx]: This could be moved to a method that's passed for CreatePassword's onCreatePassword, since that action can't fail, so you might as well call them at the same time.

This also helps you get rid of `isChangingPassword` from redux state.

File: src/app/modules/crypto/reducers.ts [hunk 6]
```diff
@@ -7,6 +7,7 @@ export interface CryptoState {
   password: null | string;
   testCipher: null | string;
   isRequestingPassword: boolean;
+  isChangingPassword: boolean;
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/reducers.ts]: You are correct, it would be ideal to have this field in the component state. I struggled with the decision to put this in redux state, but this was the only way I could get the UX that I was aiming for. The user flow is:
1. Click Change Password button
2. Prompt for current password with the `PasswordPrompt` modal (drawer remains closed)
3. After password entered, open drawer displaying the new password form
4. Enter the new password twice then click Continue button
5. Set the new password in redux & browser.storage
6. Automatically close the drawer

If we use component state, then how would the `NodeSettings` know to open the drawer at step 3? Wouldn't it have to open the drawer immediately when Change Password is clicked? Then we would also have to figure out how to inform the `NodeSettings` if the user closes the `PasswordPrompt` without entering in their current password, so that the drawer can automatically closed. Also for step 6, how can we close the drawer automatically? First thought would be to use the same `if` condition used in `componentWillUpdate` which we both agree is poo. 

I'll work on moving `isChangingPassword` to component state and try to keep the code as clean as possible. If you have any suggestions on the implementation, I am open ears.

File: src/app/modules/node/reducers.ts [hunk 7]
```diff
@@ -16,6 +16,12 @@ export interface NodeState {
   checkAuthError: null | Error;
   isFetchingNodeInfo: boolean;
   fetchNodeInfoError: null | Error;
+  
+  editingNodeField: EditingNodeField;
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/node/reducers.ts]: This field also has the same problems as `isChangingPassword` but it should be easier to move to component state because it displays the drawer first and doesn't `requirePassword` until the Continue button is clicked. I'd still have the same problem with automatically closing the drawer, but I can get around this using the condition in `componentWillUpdate`.

File: src/app/modules/crypto/sagas.ts [hunk 8]
```diff
@@ -42,7 +42,40 @@ export function* requirePassword(): SagaIterator {
...
+      yield put({
+        type: types.GENERATE_SALT,
+        payload: prevSalt,
+      })
+      return;
+    }
+
+    // clear the saves password cache since it's no longer valid
+    yield call(clearPasswordCache);
+    yield put({ type: types.CHANGE_PASSWORD_SUCCESS });
```
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/sagas.ts]: I originally had it saving to cache, then realized that we'd want to first check if a password was already cached since the user may have unchecked the `Don't ask me again this session` checkbox. I don't think its good to cache the new password if the user didn't want the old one cached. Implementing this conditional logic added a bit more code to the saga that I wasn't sure was necessary. I'll add it back and get your thoughts on whether or not it's worth it.

File: src/app/modules/node/sagas.ts (hunk 4 above)
Comment (jamaljsr) [reply src/app/modules/node/sagas.ts]: Ewwww.. great catch! Not sure how I missed that. I'll fix it.

File: src/app/modules/crypto/reducers.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (wbobeirne) [reply src/app/modules/crypto/reducers.ts]: Password prompt and salt generation could all be done _after_ the user enters their new password (CreatePassword onCreatePassword's prop action, instead of on button click.) This would also remove the need to set a new salt, and then revert if they cancel, as seen in the changePassword saga.

[the remark this answers is not in the document]
Comment (wbobeirne) [reply src/app/modules/crypto/reducers.ts]: You could auto-close drawer using a heuristic prop change (e.g. if prevProps.salt !== this.props.salt)

File: src/app/components/Settings/NodeSettings.tsx (hunk 5 above)
Comment (jamaljsr) [reply src/app/components/Settings/NodeSettings.tsx]: I'm not 100% sure I am following you here. Are you saying to create a private function in `NodeSettings`  that is assigned to `<CreatePassword onCreatePassword={this.newFunction} ...` and in this function, call both `this.props.setPassword()` and `message.success()`?

File: src/app/modules/crypto/sagas.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (wbobeirne) [reply src/app/modules/crypto/sagas.ts]: Ah, that's a good thought! Should be a pretty small implementation though,
```ts
cachedPassword = yield call(getPasswordCache);
yield put(enterPassword({
  password: newPassword,
  save: !!cachedPassword,
});
```

File: src/app/components/Settings/NodeSettings.tsx (hunk 5 above)
Comment (wbobeirne) [reply src/app/components/Settings/NodeSettings.tsx]: Yep, that's what I had in mind! Though after the change I proposed here: https://github.com/wbobeirne/joule-extension/pull/133#discussion_r247167131 you would want it in the same place as the drawer close, rather than the `onCreatePassword`, so it may have to remain in `componentWillUpdate` after all.

**Comment by jamaljsr** [comment]
@wbobeirne There wasn't a place to respond to your overall review so I'm adding a new comment. 😄 

Thanks so much for reviewing the PR and providing a lot of excellent feedback. I'm getting a better feel of your train of thought, which doesn't always translate through the code. I'll be sure to keep these things in mind for future PR's.

File: src/app/modules/crypto/reducers.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/reducers.ts]: Ah ok, I see what you are saying. But you don't think it's a bit weird to first ask a user to type in the new password twice, then to get prompted to enter their old password? The node url and macaroon forms work this way but I felt it necessary to change the flow for the password to prompt for the old one first. If you prefer to prompt after then I'll code it up this way, no prob.

**Comment by wbobeirne** [comment]
👍 thanks for bearing with me, this task is definitely more complex than average. If the community around Joule gets much bigger, I'll definitely try to put a lot of this type of feedback down in a best practices doc.

File: src/app/modules/crypto/sagas.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/sagas.ts]: Well, that's far simpler than what I had in mind 😆 . Thanks for the code snippet. I'll use that.

File: src/app/modules/crypto/reducers.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/reducers.ts]: Yup, I think this 3 field approach is definitely the best way to do this. I just have one final question about what you think is the best way to do this. I can immediately think of 3 ways to achieve what we want considering the current `CreatePassword` component needs to behave the same in the onboarding process.

1. Add a boolean `requestCurrentPassword` prop to `CreatePassword` and when set to true, it will display the current password field on top. When the submit button is clicked, the component will call `this.props.onCreatePassword` with both the current and new passwords.
2. Add the current password form field in the `NodeSettings` component and render it above the `<CreatePassword />` component in the drawer. 
3. Create a new `ChangePassword` component that would basically be a copy of `CreatePassword` but it will have the 3 fields instead of two. I could probably refactor the password strength indicator into a shared component.

In all three cases, I'd need to add the logic to test the current password against the `TEST_CIPHER_DATA`.

Please let me know what you think is best.

[the remark this answers is not in the document]
Comment (wbobeirne) [reply src/app/modules/crypto/reducers.ts]: Both 1 & 2 sound like good approaches, I'll defer to you on what you think is best between the two.

[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/reducers.ts]: I used the first approach

File: src/app/modules/node/reducers.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/node/reducers.ts]: I removed this field and put it into state. You were right about this change. The code is easier to follow.

File: src/app/modules/crypto/sagas.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (jamaljsr) [reply src/app/modules/crypto/sagas.ts]: I added this code to cache the new password

File: src/app/modules/node/sagas.ts (hunk 4 above)
Comment (jamaljsr) [reply src/app/modules/node/sagas.ts]: I fixed this by just calling `setNode` using the old url if the update fails.

**Comment by jamaljsr** [comment]
I made all of the changes requested. I still want to do a final test of all of the changes and edge cases. Feel free to provide feedback on the new commits when you have a chance. Thanks

**Comment by wbobeirne** [comment]
The code changes all look great 👌  I'll run through some manual testing tomorrow, and if it all goes well, I'll merge it in.
